nix/overlays/checked-shell-script/checked-shell-script.nix

```diff
@@ -1,17 +1,48 @@
 # Create a bash script that is checked with shellcheck. You can either use it
 # directly, or use the .bin attribute to get the script in a bin/ directory,
 # to be used in a path for example.
-{ git
+{ argbash
+, bash_5
+, git
 , lib
 , runCommand
-, runtimeShell
 , shellcheck
 , stdenv
 , writeTextFile
 }:
-{ name, docs, inRootDir ? false, redirectTixFiles ? true, withTmpDir ? false }: text:
-# TODO: do something sensible with docs, e.g. provide automated --help
+{ name
+, docs
+, args ? [ ]
+, inRootDir ? false
+, redirectTixFiles ? true
+, withEnv ? null
+, withTmpDir ? false
+}: text:
 let
+  argsParser =
+    writeTextFile {
+      name = "${name}-args";
+      destination = "/parser.sh";
+
+      text =
+        ''
+          # BASH_ARGV0 sets $0 - which is used in parser.sh for usage information
+          # stripping the /nix/store/... path for nicer display
+          BASH_ARGV0="$(basename "$0")"
+
+          # ARG_HELP([${name}], [${lib.strings.stringAsChars (c: if c == "\n" then " \\n" else c) docs}])
+          ${lib.strings.concatMapStrings (arg: "# " + arg) args}
+          # ARG_DEFAULTS_POS()
+          # ARGBASH_GO
+        '';
+
+      checkPhase =
+        ''
+          # run argbash to turn the template into a shell script
+          ${argbash}/bin/argbash $out/parser.sh -o $out/parser.sh
+        '';
```

Comment:
This is a bit of a hack, since we're modifying the written script in the `checkPhase`. Should we use a different builder?

Reply:
It would be a bit cleaner with a separate `runCommand` (we already use that in a few places). It could create a file directly without a directory, putting the output into $out. Same could be done with the textfile, no destination required.

Is there a difference between the sourcing method you chose and the `# [` and `# ]` weirdness I saw in the argbash docs?

Reply:
Just read the docs, makes sense!

Reply:
> It would be a bit cleaner with a separate `runCommand` (we already use that in a few places). It could create a file directly without a directory, putting the output into $out. Same could be done with the textfile, no destination required.

So you suggest one textfile `argTemplate` and a runCommand `argParser` or something? We could also make it one runCommand and pipe the template into stdin of `argbash`. WDYT?

> Is there a difference between the sourcing method you chose and the `# [` and `# ]` weirdness I saw in the argbash docs?

That weirdness is about the ability to run argbash against the output file **again**, e.g. after modification. The sourcing method does play a small role, but not that much. Since we're starting from scratch all the time, due to how nix works, we don't need to worry about that.

Reply:
Well need the template for the completion also, think it will be good to have it separare!

Reply:
Hm. I think I'd run the completion command on the final script, not the "to-be-sourced-args-script", because from [the docs](https://argbash.readthedocs.io/en/latest/usage.html#bash-completion):

> Notice that in this case, the completion file has to “know” the basename of the script the completion is meant for. The basename is inferred either from the source filename, or from the destination by stripping the .m4 suffix if applicable

Reply:
Hm, not sure - will be best to try it out. It will need some version of the file with the macros in it, no?

> So you suggest one textfile argTemplate and a runCommand argParser or something? We could also make it one runCommand and pipe the template into stdin of argbash. WDYT?

I'd go with the first, but both is fine!

Reply:
> Hm, not sure - will be best to try it out. It will need some version of the file with the macros in it, no?

Yes, but the macros are copied to the final script, too.

I'm currently looking at running `argbash --type complete` in the `script` phase of checked-shell-script, i.e. where the symbolic link is created - and I wonder, what the reason is to create that symbolic link in the first place? Currently this just creates another `.drv` files to add the link - why are we not executing the script directly?

Reply:
We are returning two versions of the script at the end: `script // { inherit bin; }`

`script` is easier to use within our scripts, as you can just go `${script}` instead of `${script}/bin/script`. `bin` has the same script in a `bin` subdirectory, which is needed to put our scripts on PATHs. We access that attribute to use it.

When we add bash completion, I would imagine that we have an additional `bashCompletion` attribute (as we do e.g. for `hsie`). The last line will then be:

```
`script // { inherit bin bashCompletion; }`
```

Reply:
> We are returning two versions of the script at the end: `script // { inherit bin; }`
> 
> `script` is easier to use within our scripts, as you can just go `${script}` instead of `${script}/bin/script`. `bin` has the same script in a `bin` subdirectory, which is needed to put our scripts on PATHs. We access that attribute to use it.

Ah! That's where the magic comes from. Ok, got it.

> When we add bash completion, I would imagine that we have an additional `bashCompletion` attribute (as we do e.g. for `hsie`).

Yes, I came to that conclusion, too. Now, I just need to export all those to `shell.nix`. I think I will create a single completion script for each "module" (devtools, style, ...), which sources all the tools. This would also to move the `_command` completions we currently have in `shell.nix` closer to their shell-scripts.

Reply:
> > Hm, not sure - will be best to try it out. It will need some version of the file with the macros in it, no?
> 
> Yes, but the macros are copied to the final script, too.

But we're sourcing that part only... argh! I see the problem now ;)